server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridTiler.java

```diff
@@ -0,0 +1,256 @@
+                values.resizeCell(1);
+                values.add(0, GeoTileUtils.longEncodeTiles(precision, minXTile, minYTile));
+                return 1;
+            } else if (count <= precision) {
+                return setValuesByBruteForceScan(values, geoValue, precision, minXTile, minYTile, maxXTile, maxYTile);
+            } else {
+                return setValuesByRasterization(0, 0, 0, values, 0, precision, geoValue, bounds);
+            }
+        }
+
+        /**
+         *
+         * @param values the bucket values as longs
+         * @param geoValue the shape value
+         * @param precision the target precision to split the shape up into
+         * @return the number of buckets the geoValue is found in
+         */
+        protected int setValuesByBruteForceScan(CellIdSource.GeoShapeCellValues values, MultiGeoValues.GeoValue geoValue,
+                                                int precision, int minXTile, int minYTile, int maxXTile, int maxYTile) {
+            int idx = 0;
+            for (int i = minXTile; i <= maxXTile; i++) {
+                for (int j = minYTile; j <= maxYTile; j++) {
+                    Rectangle rectangle = GeoTileUtils.toBoundingBox(i, j, precision);
+                    if (geoValue.relate(rectangle) != GeoRelation.QUERY_DISJOINT) {
+                        values.resizeCell(idx + 1);
+                        values.add(idx++, GeoTileUtils.longEncodeTiles(precision, i, j));
+                    }
+                }
+            }
+            return idx;
+        }
+
+        protected int setValuesByRasterization(int xTile, int yTile, int zTile, CellIdSource.GeoShapeCellValues values,
+                                               int valuesIndex, int targetPrecision, MultiGeoValues.GeoValue geoValue,
+                                               MultiGeoValues.BoundingBox shapeBounds) {
+            zTile++;
+            for (int i = 0; i < 2; i++) {
+                for (int j = 0; j < 2; j++) {
+                    int nextX = 2 * xTile + i;
+                    int nextY = 2 * yTile + j;
+                    Rectangle rectangle = GeoTileUtils.toBoundingBox(nextX, nextY, zTile);
+                    // TODO: this looks hacky, maybe the relate method should handle it?
+                    if (shapeBounds.minX() == rectangle.getMaxX() ||
+                        shapeBounds.maxY() == rectangle.getMinY()) {
+                        continue;
+                    }
+                    GeoRelation relation = geoValue.relate(rectangle);
+                    if (GeoRelation.QUERY_INSIDE == relation) {
+                        if (zTile == targetPrecision) {
+                            values.resizeCell(valuesIndex + 1);
+                            values.add(valuesIndex++, GeoTileUtils.longEncodeTiles(zTile, nextX, nextY));
+                        } else {
+                            values.resizeCell(valuesIndex + (int) Math.pow(4, targetPrecision - zTile) + 1);
```

Comment:
`Math#pow` is quite slow, I believe we could do this instead:
```suggestion
                            values.resizeCell(valuesIndex +  1 << ( 2 * (targetPrecision - zTile)) + 1);
```